src/libraries/Common/src/System/Net/Http/aspnetcore/Http2/Hpack/HPackDecoder.cs

```diff
@@ -124,107 +122,270 @@ public void Decode(in ReadOnlySequence<byte> data, bool endHeaders, IHttpHeaders
+                        break;
+                    case State.HeaderNameLengthContinue:
+                        ParseHeaderNameLengthContinue(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderName:
+                        ParseHeaderName(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderValueLength:
+                        ParseHeaderValueLength(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderValueLengthContinue:
+                        ParseHeaderValueLengthContinue(data, ref currentIndex, handler);
+                        break;
+                    case State.HeaderValue:
+                        ParseHeaderValue(data, ref currentIndex, handler);
+                        break;
+                    case State.DynamicTableSizeUpdate:
+                        ParseDynamicTableSizeUpdate(data, ref currentIndex);
+                        break;
+                    default:
+                        // Can't happen
+                        Debug.Fail("HPACK decoder reach an invalid state");
+                        throw new NotImplementedException(_state.ToString());
+                }
+            }
+            while (currentIndex < data.Length);
 
-                            int val = b & ~IndexedHeaderFieldMask;
+            // If a header range was set, but the value was not in the data, then copy the range
+            // to the name buffer. Must copy because because the data will be replaced and the range
+            // will no longer be valid.
+            if (!_headerNameRange.Equals(default))
+            {
+                _headerName = _headerNameOctets;
 
-                            if (_integerDecoder.BeginTryDecode((byte)val, IndexedHeaderFieldPrefix, out intResult))
-                            {
-                                OnIndexedHeaderField(intResult, handler);
-                            }
-                            else
-                            {
-                                _state = State.HeaderFieldIndex;
-                            }
-                        }
-                        else if ((b & LiteralHeaderFieldWithIncrementalIndexingMask) == LiteralHeaderFieldWithIncrementalIndexingRepresentation)
-                        {
-                            _headersObserved = true;
+                var headerBytes = data[_headerNameRange];
+                headerBytes.CopyTo(_headerName);
+                _headerNameLength = headerBytes.Length;
+                _headerNameRange = default;
+            }
+        }
 
-                            _index = true;
-                            int val = b & ~LiteralHeaderFieldWithIncrementalIndexingMask;
+        private void ParseDynamicTableSizeUpdate(ReadOnlySpan<byte> data, ref int currentIndex)
+        {
+            for (; currentIndex < data.Length; currentIndex++)
```

Comment:
This pattern -- i.e. loop over remaining bytes, calling _integerDecoder.TryDecode -- is repeated in several of the parse routines. I wonder if it's worth moving into a helper routine.